File: site/tools/generate-docs/enum.ts
import { Package } from './package';
import { JsonEnum } from './types';

const { log } = console;

export default class Enum {
  package: Package;
  fileName: string;

  name: string;
  longName: string;
  fullName: string;
  description: string;
  values: {
    name: string;
    number: string;
    description: string;
  }[];

  constructor(json: JsonEnum, pkg: Package, fileName: string) {
    log(`Creating enum ${json.name} with ${json.values.length} values`);
    this.package = pkg;
    this.fileName = fileName;
    this.name = json.name;
    this.longName = json.longName;
    this.fullName = json.fullName;
    this.description = json.description
      .replace(/\n/g, ' ') // replace newlines with spaces
      .replace(/\|/g, '\\|'); // escape pipe "|" chars
    this.values = json.values.map((v) => ({
      name: v.name,
      number: v.number,
      description: v.description
        .replace(/\n/g, ' ') // replace newlines with spaces
        .replace(/\|/g, '\\|'), // escape pipe "|" chars
    }));
  }
}

File: models/enum.go
package models

import (
	"fmt"

	"github.com/lightninglabs/lightning-api-ng/markdown"
	gendoc "github.com/pseudomuto/protoc-gen-doc"
)

type Enum struct {
	pkg      *Package
	FileName string

	Name        string
	LongName    string
	FullName    string
	Description string
	Values      []*EnumValue
}

type EnumValue struct {
	Name        string
	Number      string
	Description string
}

// NewEnum creates a new Enum model from an Enum definition
func NewEnum(enumDef *gendoc.Enum, pkg *Package, fileName string) *Enum {
	fmt.Printf("Creating enum %s with %d values\n", enumDef.FullName,
		len(enumDef.Values))

	enum := &Enum{
		pkg:      pkg,
		FileName: fileName,
		Name:     enumDef.Name,
		LongName: enumDef.LongName,
		FullName: enumDef.FullName,
		Description: markdown.CleanDescription(
			enumDef.Description, true),

		Values: make([]*EnumValue, 0),
	}
	for _, valueDef := range enumDef.Values {
		value := &EnumValue{
			Name:   valueDef.Name,
			Number: valueDef.Number,
			Description: markdown.CleanDescription(
				valueDef.Description, true),
		}
		enum.Values = append(enum.Values, value)
	}

	return enum
}
